nrf/modules/machine/pin: Add support for IRQ on Pin's

glennrub opened the pull request:
This patch ads irq method to the pin object. Handlers
registered in the irq method will be kept as part of the
ROOT_POINTERS.

In order to resolve which pin object is the root of the
IRQ, the pin_find has been extended to also be able to
search up Pin objects based on mp_int_t pin number.
This also implies that the Pin.new API is now also supporting
creation of Pin objects based on the integer value of the
pin instead of old style mandating string name of the Pin.

All boards have been updated to use real pin number from
0-48 instead of pin_Pxx for UART/SPI and music module pins.

UART/SPI/modmusic has also been updated to use pin number
provided directly or look up the Pin object based on the
integer value of the pin (modmusic).

Pin generation has been updated to create a list of pins, where
the board/cpu dicts are now refering to an index in this list
instead of having one const declaration for each pin. This new
const table makes it possible to iterate through all pins generated
in order to locate the correct Pin object.

glennrub committed:
nrf/modules/machine/pin: Add support for IRQ on Pin's

This patch ads irq method to the pin object. Handlers
registered in the irq method will be kept as part of the
ROOT_POINTERS.

In order to resolve which pin object is the root of the
IRQ, the pin_find has been extended to also be able to
search up Pin objects based on mp_int_t pin number.
This also implies that the Pin.new API is now also supporting
creation of Pin objects based on the integer value of the
pin instead of old style mandating string name of the Pin.

All boards have been updated to use real pin number from
0-48 instead of pin_Pxx for UART/SPI and music module pins.

UART/SPI/modmusic has also been updated to use pin number
provided directly or look up the Pin object based on the
integer value of the pin (modmusic).

Pin generation has been updated to create a list of pins, where
the board/cpu dicts are now refering to an index in this list
instead of having one const declaration for each pin. This new
const table makes it possible to iterate through all pins generated
in order to locate the correct Pin object.

glennrub commented:
Need to look into this one a bit more. There are issues using the IRQ in combination with Bluetooth LE. If a print is performed in callback function, it looks like the BLE REPL gets stuck after first characteristic write. Also, priority level for nrf51 is wrong in the nrfx_config.h.

aykevl commented:
Maybe the SD event handler (SWI2 or something like it) has the same priority? The pin IRQ must be lower for NUS to work.

glennrub commented:
yeah, there must be something like that.

Currently i see that it gets stuck in ble_drv.c:671
```
void ble_drv_attr_s_notify(uint16_t conn_handle, uint16_t handle, uint16_t len, uint8_t * p_data) {
...
    while (m_tx_in_progress) {
        ;
    }
...
}
```
I assume the pin IRQ is controlled by the NRFX_GPIOTE_CONFIG_IRQ_PRIORITY, which i have set ot both 6 and 7, without any change in behavior. According to SoftDevice specification for s132 3.0.0 the SVC is running in priority 4 and should have unlocked this loop. 
```

glennrub commented:
@aykevl does this look like it could be the issue you are solving in PR #157 ?

aykevl commented:
No, #157 shouldn't solve it. It only turns a bool into an int. But it will probably show a few more characters before deadlocking.

You could maybe try other interrupt types, like RTC?

Oh and how are you exactly printing? Note that the GC/allocator may be locked during an IRQ.

glennrub commented:
After investigating the issues, I believe the issues related to getting a functional IRQ handler while using BLE REPL is more a ble_drv.c issues and not really related to Pin IRQ level. Adding PR #157 helps on nrf52 target, while nrf51 BLE REPL seems to have other issues. The recommendation for now would be to not do long print's in the Pin handler (< 20 bytes).

Size wise this patch adds about 1.5K flash size because of the gpiote driver and additional glue code in pin.c.

aykevl commented on ports/nrf/modules/machine/pin.c:
```diff
@@ -105,6 +113,12 @@ STATIC bool pin_class_debug;
+    MP_STATE_PORT(machine_pin_irq_handlers) = &irq_handlers[0];
+
+    // Initialize GPIOTE if not done yet.
+    if (nrfx_gpiote_is_init() == false) {
```
1. Why not just `if (!nrfx_gpiote_is_init())`
2. I think you can skip this check entirely - it is already checked at initialization (but logs an error if enable)

glennrub replied:
1. I got a compiler warning telling me that the return value was not used. But it seems to have gone away now. I'll efix this.
2. If CTRL+D is hit (soft-reset) this module is not re-initilalized giving errors. The same applies to registered pins.

aykevl commented on ports/nrf/mpconfigport.h:
```diff
@@ -312,6 +312,7 @@ extern const struct _mp_obj_module_t ble_module;
     const char *readline_hist[8]; \
     mp_obj_t pin_class_mapper; \
     mp_obj_t pin_class_map_dict; \
+    struct _machine_pin_irq_handler_t * machine_pin_irq_handlers; \
```
You could move the entire `irq_handlers` over to here I think, just like `readline_hist` for example.

glennrub replied:
The reason why i put it in pin.c is that i had a very hard time resolving the NUMBER_OF_PINS from the include hierarchy of nrfx.h inside this header. i'll give it a second try.

aykevl replied:
This is actually required: the heap allocated pointers must be directly stored in here otherwise the GC won't pick up the root pointers. If it is done via an indirect (.bss) pointer, the GC will consider the pointer to be outside the heap and won't scan it any further.

glennrub commented:
@aykevl, I've pushed a forced update. Could you take a second look if it looks a bit more like what you expected?

glennrub commented:
Updated code with force push, and also added a description text of this PR, taken from the updated commit message. I still lack to test how the integration went with modmusic and pin lookup.
